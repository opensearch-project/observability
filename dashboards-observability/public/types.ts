/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DashboardStart } from '../../../src/plugins/dashboard/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dashboard: DashboardStart;
}

export interface ObservabilitySetup {}

export interface ObservabilityStart {}
