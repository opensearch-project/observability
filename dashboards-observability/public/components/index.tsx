/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { QueryManager } from 'common/query_manager';
import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { AppPluginStartDependencies } from '../types';
import { App } from './app';

export const Observability = (
  CoreStartProp: CoreStart,
  DepsStart: AppPluginStartDependencies,
  AppMountParametersProp: AppMountParameters,
  pplService: any,
  dslService: any,
  savedObjects: any,
  timestampUtils: any,
  queryManager: QueryManager
) => {
  ReactDOM.render(
    <App
      CoreStartProp={CoreStartProp}
      DepsStart={DepsStart}
      pplService={pplService}
      dslService={dslService}
      savedObjects={savedObjects}
      timestampUtils={timestampUtils}
      queryManager={queryManager}
    />,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
