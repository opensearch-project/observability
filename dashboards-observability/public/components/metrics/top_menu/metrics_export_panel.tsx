/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CUSTOM_PANELS_API_PREFIX } from '../../../../common/constants/custom_panels';
import React, { useEffect, useState } from 'react';
import { CoreStart } from '../../../../../../src/core/public';
import {
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFieldText,
  EuiFlexGroup,
  EuiFormRow,
  EuiFlexItem,
  EuiForm,
} from '@elastic/eui';
import { createPrometheusMetricById } from '../helpers/utils';
import { MetricType } from '../../../../common/types/metrics';
import { fetchVisualizationById } from '../../custom_panels/helpers/utils';

interface MetricsExportPanelProps {
  http: CoreStart['http'];
  visualizationsMetaData: any;
  setVisualizationsMetaData: React.Dispatch<any>;
  sortedMetricsLayout: MetricType[];
  selectedPanelOptions: EuiComboBoxOptionOption<unknown>[] | undefined;
  setSelectedPanelOptions: React.Dispatch<
    React.SetStateAction<EuiComboBoxOptionOption<unknown>[] | undefined>
  >;
}

interface CustomPanelOptions {
  id: string;
  name: string;
  dateCreated: string;
  dateModified: string;
}

export const MetricsExportPanel = ({
  http,
  visualizationsMetaData,
  setVisualizationsMetaData,
  sortedMetricsLayout,
  selectedPanelOptions,
  setSelectedPanelOptions,
}: MetricsExportPanelProps) => {
  const [options, setOptions] = useState([]);

  const [errorResponse, setErrorResponse] = useState('');

  const getCustomPanelList = async () => {
    http
      .get(`${CUSTOM_PANELS_API_PREFIX}/panels`)
      .then((res: any) => {
        setOptions(res.panels || []);
      })
      .catch((error: any) => console.error(error));
  };

  const fetchAllvisualizationsById = async () => {
    let tempVisualizationsMetaData = await Promise.all(
      sortedMetricsLayout.map(async (metricLayout) => {
        return metricLayout.metricType === 'savedCustomMetric'
          ? await fetchVisualizationById(http, metricLayout.id, setErrorResponse)
          : createPrometheusMetricById(metricLayout.id);
      })
    );
    setVisualizationsMetaData(tempVisualizationsMetaData);
  };

  useEffect(() => {
    getCustomPanelList();
    fetchAllvisualizationsById();
  }, []);

  const onNameChange = (index: number, name: string) => {
    let tempVisualizationsMetaData = [...visualizationsMetaData];
    tempVisualizationsMetaData[index].name = name;
    setVisualizationsMetaData(tempVisualizationsMetaData);
  };

  return (
    <div style={{ minWidth: '15vw' }}>
      <EuiFormRow
        label="Custom operational dashboards/application"
        helpText="Search existing dashboards or applications by name"
      >
        <EuiComboBox
          placeholder="Select dashboards/applications"
          onChange={(options) => {
            setSelectedPanelOptions(options);
          }}
          selectedOptions={selectedPanelOptions}
          options={options.map((option: CustomPanelOptions) => {
            return {
              panel: option,
              label: option.name,
            };
          })}
          isClearable={true}
          data-test-subj="eventExplorer__querySaveComboBox"
        />
      </EuiFormRow>

      {visualizationsMetaData.length > 0 && (
        <div style={{ maxHeight: '30vh', overflowY: 'scroll', width: 'auto', overflowX: 'hidden' }}>
          {visualizationsMetaData.map((metaData: any, index: number) => {
            return (
              <EuiForm component="form">
                <EuiFlexGroup>
                  <EuiFlexItem>
                    <EuiFormRow label={'Metric Name #' + (index + 1)}>
                      <EuiFieldText
                        key={'save-panel-id'}
                        value={visualizationsMetaData[index].name}
                        onChange={(e) => onNameChange(index, e.target.value)}
                        data-test-subj="metrics__querySaveName"
                      />
                    </EuiFormRow>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </EuiForm>
            );
          })}
        </div>
      )}
    </div>
  );
};
