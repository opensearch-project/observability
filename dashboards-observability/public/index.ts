/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './components/trace_analytics/index.scss';
import './components/notebooks/index.scss'

import { PluginInitializer, PluginInitializerContext } from '../../../src/core/public';
import {
  ObservabilityPlugin
} from './plugin';

export { ObservabilityPlugin as Plugin };

export const plugin = (initializerContext: PluginInitializerContext) => new ObservabilityPlugin(initializerContext);