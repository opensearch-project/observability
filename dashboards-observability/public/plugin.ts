/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../src/core/public';
import {
  observabilityID,
  observabilityPluginOrder,
  observabilityTitle,
} from '../common/constants/shared';
import PPLService from './services/requests/ppl';
import DSLService from './services/requests/dsl';
import TimestampUtils from './services/timestamp/timestamp';
import SavedObjects from './services/saved_objects/event_analytics/saved_objects';
import { AppPluginStartDependencies, ObservabilitySetup, ObservabilityStart } from './types';
import { convertLegacyNotebooksUrl } from './components/notebooks/components/helpers/legacy_route_helpers';
import { convertLegacyTraceAnalyticsUrl } from './components/trace_analytics/components/common/legacy_route_helpers';
import { uiSettingsService } from '../common/utils';
import { QueryManager } from '../common/query_manager';
export class ObservabilityPlugin implements Plugin<ObservabilitySetup, ObservabilityStart> {
  public setup(core: CoreSetup): ObservabilitySetup {
    uiSettingsService.init(core.uiSettings, core.notifications);

    // redirect legacy notebooks URL to current URL under observability
    if (window.location.pathname.includes('notebooks-dashboards')) {
      window.location.assign(convertLegacyNotebooksUrl(window.location));
    }

    // redirect legacy trace analytics URL to current URL under observability
    if (window.location.pathname.includes('trace-analytics-dashboards')) {
      window.location.assign(convertLegacyTraceAnalyticsUrl(window.location));
    }

    core.application.register({
      id: observabilityID,
      title: observabilityTitle,
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: observabilityPluginOrder,
      async mount(params: AppMountParameters) {
        const { Observability } = await import('./components/index');
        const [coreStart, depsStart] = await core.getStartServices();
        const pplService = new PPLService(coreStart.http);
        const dslService = new DSLService(coreStart.http);
        const savedObjects = new SavedObjects(coreStart.http);
        const timestampUtils = new TimestampUtils(dslService, pplService);
        const qm = new QueryManager();
        return Observability(
          coreStart,
          depsStart as AppPluginStartDependencies,
          params,
          pplService,
          dslService,
          savedObjects,
          timestampUtils,
          qm
        );
      },
    });

    // Return methods that should be available to other plugins
    return {};
  }
  public start(core: CoreStart): ObservabilityStart {
    return {};
  }
  public stop() {}
}
