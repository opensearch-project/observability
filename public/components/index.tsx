/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { AppPluginStartDependencies } from '../types';
import { App } from './app';

export const Observability = (
  CoreStart: CoreStart,
  DepsStart: AppPluginStartDependencies,
  AppMountParameters: AppMountParameters,
  pplService: any,
  dslService: any,
  savedObjects: any
) => {
  ReactDOM.render(
    <App
      CoreStart={ CoreStart }
      DepsStart={DepsStart} 
      pplService={ pplService }
      dslService ={ dslService }
      savedObjects={ savedObjects }
    />,
    AppMountParameters.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParameters.element);
};
