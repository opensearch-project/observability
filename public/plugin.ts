/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../src/core/public';
import {
  observabilityID,
  observabilityPluginOrder,
  observabilityTitle,
} from '../common/constants/shared';
import PPLService from './services/requests/ppl';
import DSLService from './services/requests/dsl';
import SavedObjects from './services/saved_objects/event_analytics/saved_objects';
import { AppPluginStartDependencies, ObservabilitySetup, ObservabilityStart } from './types';

export class ObservabilityPlugin implements Plugin<ObservabilitySetup, ObservabilityStart> {

    public setup(core: CoreSetup): ObservabilitySetup {
      
      core.application.register({
        id: observabilityID,
        title: observabilityTitle,
        category: {
          id: 'opensearch',
          label: 'OpenSearch Plugins',
          order: 2000,
        },
        order: observabilityPluginOrder,
        async mount(params: AppMountParameters) {
          const { Observability } = await import('./components/index');
          const [ coreStart, depsStart ] = await core.getStartServices();
          const pplService = new PPLService(coreStart.http);
          const dslService = new DSLService(coreStart.http);
          const savedObjects = new SavedObjects(coreStart.http);
          return Observability(
            coreStart,
            depsStart as AppPluginStartDependencies,
            params,
            pplService,
            dslService,
            savedObjects
          );
        },
      });

      // Return methods that should be available to other plugins
      return {};
    }
    public start(core: CoreStart): ObservabilityStart {
      return {};
    }
    public stop() {}
}
