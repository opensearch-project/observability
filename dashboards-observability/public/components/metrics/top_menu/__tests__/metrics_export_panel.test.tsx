/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import httpClientMock from '../../../../../test/__mocks__/httpClientMock';
import {
  samplePanelOptions,
  sampleSortedMetricsLayout,
  sampleVisualizationById,
} from '../../../../../test/metrics_contants';
import { createStore } from '@reduxjs/toolkit';
import rootReducer from '../../../../framework/redux/reducers';
import { Provider } from 'react-redux';
import { HttpResponse } from '../../../../../../../src/core/public';
import { MetricsExportPanel } from '../metrics_export_panel';
import { EuiComboBoxOptionOption } from '@elastic/eui';

describe('Export Metrics Panel Component', () => {
  configure({ adapter: new Adapter() });
  const store = createStore(rootReducer);

  it('renders Export Metrics Panel Component', async () => {
    let httpFlag = 1;
    httpClientMock.get = jest.fn(() => {
      if (httpFlag == 1) {
        httpFlag += 1;
        return Promise.resolve((samplePanelOptions as unknown) as HttpResponse);
      } else {
        return Promise.resolve((sampleVisualizationById as unknown) as HttpResponse);
      }
    });

    const http = httpClientMock;
    const visualizationsMetaData: any = [];
    const setVisualizationsMetaData = jest.fn();
    const sortedMetricsLayout = sampleSortedMetricsLayout;
    const selectedPanelOptions: EuiComboBoxOptionOption<unknown>[] = [];
    const setSelectedPanelOptions = jest.fn();

    const wrapper = mount(
      <Provider store={store}>
        <MetricsExportPanel
          http={http}
          visualizationsMetaData={visualizationsMetaData}
          setVisualizationsMetaData={setVisualizationsMetaData}
          sortedMetricsLayout={sortedMetricsLayout}
          selectedPanelOptions={selectedPanelOptions}
          setSelectedPanelOptions={setSelectedPanelOptions}
        />
      </Provider>
    );
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
