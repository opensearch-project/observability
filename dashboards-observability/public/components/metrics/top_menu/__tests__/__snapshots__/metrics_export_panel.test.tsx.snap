// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Metrics Panel Component renders Export Metrics Panel Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <MetricsExportPanel
    http={[MockFunction]}
    selectedPanelOptions={Array []}
    setSelectedPanelOptions={[MockFunction]}
    setVisualizationsMetaData={[MockFunction]}
    sortedMetricsLayout={
      Array [
        Object {
          "h": 2,
          "id": "Y4muP4QBiaYaSxpXk7r8",
          "metricType": "savedCustomMetric",
          "savedVisualizationId": "Y4muP4QBiaYaSxpXk7r8",
          "w": 12,
          "x": 0,
          "y": 0,
        },
        Object {
          "h": 2,
          "id": "prometheus.process_resident_memory_bytes",
          "metricType": "prometheusMetric",
          "savedVisualizationId": "prometheus.process_resident_memory_bytes",
          "w": 12,
          "x": 0,
          "y": 2,
        },
      ]
    }
    visualizationsMetaData={Array []}
  >
    <div
      style={
        Object {
          "minWidth": "15vw",
        }
      }
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        helpText="Search existing dashboards or applications by name"
        label="Custom operational dashboards/application"
        labelType="label"
      >
        <div
          className="euiFormRow"
          id="random_html_id-row"
        >
          <div
            className="euiFormRow__labelWrapper"
          >
            <EuiFormLabel
              className="euiFormRow__label"
              htmlFor="random_html_id"
              isFocused={false}
              type="label"
            >
              <label
                className="euiFormLabel euiFormRow__label"
                htmlFor="random_html_id"
              >
                Custom operational dashboards/application
              </label>
            </EuiFormLabel>
          </div>
          <div
            className="euiFormRow__fieldWrapper"
          >
            <EuiComboBox
              aria-describedby="random_html_id-help-0"
              async={false}
              compressed={false}
              data-test-subj="eventExplorer__querySaveComboBox"
              fullWidth={false}
              id="random_html_id"
              isClearable={true}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              options={Array []}
              placeholder="Select dashboards/applications"
              selectedOptions={Array []}
              singleSelection={false}
              sortMatchesBy="none"
            >
              <div
                aria-describedby="random_html_id-help-0"
                aria-expanded={false}
                aria-haspopup="listbox"
                className="euiComboBox"
                data-test-subj="eventExplorer__querySaveComboBox"
                onFocus={[Function]}
                onKeyDown={[Function]}
                role="combobox"
              >
                <EuiComboBoxInput
                  autoSizeInputRef={[Function]}
                  compressed={false}
                  fullWidth={false}
                  hasSelectedOptions={false}
                  id="random_html_id"
                  inputRef={[Function]}
                  isListOpen={false}
                  noIcon={false}
                  onChange={[Function]}
                  onClear={[Function]}
                  onClick={[Function]}
                  onCloseListClick={[Function]}
                  onFocus={[Function]}
                  onOpenListClick={[Function]}
                  onRemoveOption={[Function]}
                  placeholder="Select dashboards/applications"
                  rootId={[Function]}
                  searchValue=""
                  selectedOptions={Array []}
                  singleSelection={false}
                  toggleButtonRef={[Function]}
                  updatePosition={[Function]}
                  value=""
                >
                  <EuiFormControlLayout
                    compressed={false}
                    fullWidth={false}
                    icon={
                      Object {
                        "aria-label": "Open list of options",
                        "data-test-subj": "comboBoxToggleListButton",
                        "disabled": undefined,
                        "onClick": [Function],
                        "ref": [Function],
                        "side": "right",
                        "type": "arrowDown",
                      }
                    }
                  >
                    <div
                      className="euiFormControlLayout"
                    >
                      <div
                        className="euiFormControlLayout__childrenWrapper"
                      >
                        <div
                          className="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
                          data-test-subj="comboBoxInput"
                          onClick={[Function]}
                          tabIndex={-1}
                        >
                          <p
                            className="euiComboBoxPlaceholder"
                          >
                            Select dashboards/applications
                          </p>
                          <AutosizeInput
                            aria-controls=""
                            className="euiComboBox__input"
                            data-test-subj="comboBoxSearchInput"
                            id="random_html_id"
                            injectStyles={true}
                            inputRef={[Function]}
                            minWidth={1}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            role="textbox"
                            style={
                              Object {
                                "fontSize": 14,
                              }
                            }
                            value=""
                          >
                            <div
                              className="euiComboBox__input"
                              style={
                                Object {
                                  "display": "inline-block",
                                  "fontSize": 14,
                                }
                              }
                            >
                              <input
                                aria-controls=""
                                data-test-subj="comboBoxSearchInput"
                                id="random_html_id"
                                onBlur={[Function]}
                                onChange={[Function]}
                                onFocus={[Function]}
                                role="textbox"
                                style={
                                  Object {
                                    "boxSizing": "content-box",
                                    "width": "2px",
                                  }
                                }
                                value=""
                              />
                              <div
                                style={
                                  Object {
                                    "height": 0,
                                    "left": 0,
                                    "overflow": "scroll",
                                    "position": "absolute",
                                    "top": 0,
                                    "visibility": "hidden",
                                    "whiteSpace": "pre",
                                  }
                                }
                              />
                            </div>
                          </AutosizeInput>
                        </div>
                        <EuiFormControlLayoutIcons
                          compressed={false}
                          icon={
                            Object {
                              "aria-label": "Open list of options",
                              "data-test-subj": "comboBoxToggleListButton",
                              "disabled": undefined,
                              "onClick": [Function],
                              "ref": [Function],
                              "side": "right",
                              "type": "arrowDown",
                            }
                          }
                        >
                          <div
                            className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                          >
                            <EuiFormControlLayoutCustomIcon
                              aria-label="Open list of options"
                              data-test-subj="comboBoxToggleListButton"
                              iconRef={[Function]}
                              onClick={[Function]}
                              size="m"
                              type="arrowDown"
                            >
                              <button
                                aria-label="Open list of options"
                                className="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                                data-test-subj="comboBoxToggleListButton"
                                onClick={[Function]}
                                type="button"
                              >
                                <EuiIcon
                                  aria-hidden="true"
                                  className="euiFormControlLayoutCustomIcon__icon"
                                  size="m"
                                  type="arrowDown"
                                >
                                  <EuiIconEmpty
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    />
                                  </EuiIconEmpty>
                                </EuiIcon>
                              </button>
                            </EuiFormControlLayoutCustomIcon>
                          </div>
                        </EuiFormControlLayoutIcons>
                      </div>
                    </div>
                  </EuiFormControlLayout>
                </EuiComboBoxInput>
              </div>
            </EuiComboBox>
            <EuiFormHelpText
              className="euiFormRow__text"
              id="random_html_id-help-0"
              key="Search existing dashboards or applications by name"
            >
              <div
                className="euiFormHelpText euiFormRow__text"
                id="random_html_id-help-0"
              >
                Search existing dashboards or applications by name
              </div>
            </EuiFormHelpText>
          </div>
        </div>
      </EuiFormRow>
    </div>
  </MetricsExportPanel>
</Provider>
`;
